As a user, if I declare a WRDS threshold feature_name_from of "baseline" but have declared no baseline, I want to see a meaningful validation error instead of an NPE

---


Author Name: **Hank** (Hank)
Original Redmine Issue: 99409, https://vlab.noaa.gov/redmine/issues/99409
Original Date: 2021-12-08


---

Example declaration will be provided in the first comment. 

The subject is pretty clear.  This ticket can be resolved when we validate against a featureNameFrom that points to baseline with no baseline source provided.  

This might be a low hanging fruit.  I'll say 4 hours for now, but feel that may be high.

Hank

---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2021-12-08T14:38:58Z

---

Declaration:

```<code class="xml">
<?xml version='1.0' encoding='UTF-8'?>
<project>
  <inputs>
    <left>
      <type>observations</type>
      <source interface="usgs_nwis">https://nwis.waterservices.usgs.gov/nwis/iv</source>
      <variable>00060</variable>
    </left>
        <right label="NWM Medium Range Single Valued">
            <type>single valued forecasts</type>
            <source interface="nwm_medium_range_deterministic_channel_rt_conus_hourly">https://nwcal-dstore.[host]/nwm/2.1/</source>
            <variable>streamflow</variable>
        </right>
  </inputs>
  <pair>
    <unit>ft3/s</unit>
    <featureService>
      <baseUrl>https://nwcal-wrds.[host]/api/location/v3.0/metadata</baseUrl>
      <group pool="true">
        <type>state</type>
        <value>NJ</value>
      </group>
    </featureService>

    <leadHours minimum="0" maximum="240"/>
    <dates earliest="2021-08-24T00:00:00Z" latest="2021-09-08T00:00:00Z"/>
    <issuedDates earliest="2021-08-24T00:00:00Z" latest="2021-09-08T00:00:00Z"/>
    <leadTimesPoolingWindow>
      <period>3</period>
      <unit>hours</unit>
    </leadTimesPoolingWindow>
  </pair>
  <metrics>
    <thresholds>
      <type>value</type>
      <applyTo>left</applyTo>
      <source format="wrds" parameterToMeasure="flow" featureNameFrom="baseline" >https://nwcal-wrds.[host]/api/location/v3.0/nws_threshold/nws_lid</source>
      <operator>greater than or equal to</operator>
    </thresholds>

    <metric><name>sample size</name></metric>
    <metric><name>pearson correlation coefficient</name></metric>
  </metrics>
  <outputs>
    <destination type="graphic" />
    <destination type="csv" />
    <destination type="csv2" />
    <destination type="pairs"/>
  </outputs>
</project>
</code>
```

Complete exception:

```
javax.jms.IllegalStateException: Session is already closed
        at org.apache.qpid.client.AMQSession.acknowledge(AMQSession.java:526)
        at org.apache.qpid.client.message.AbstractAMQMessageDelegate.acknowledge(AbstractAMQMessageDelegate.java:219)
        at org.apache.qpid.client.message.AbstractJMSMessage.acknowledge(AbstractJMSMessage.java:311)
        at wres.events.subscribe.EvaluationSubscriber.lambda$getEvaluationListener$3(EvaluationSubscriber.java:916)
        at org.apache.qpid.client.BasicMessageConsumer.notifyMessage(BasicMessageConsumer.java:758)
        at org.apache.qpid.client.BasicMessageConsumer.notifyMessage(BasicMessageConsumer.java:732)
        at org.apache.qpid.client.AMQSession$Dispatcher.notifyConsumer(AMQSession.java:3705)
        at org.apache.qpid.client.AMQSession$Dispatcher.dispatchMessage(AMQSession.java:3637)
        at org.apache.qpid.client.AMQSession$Dispatcher.access$1300(AMQSession.java:3406)
        at org.apache.qpid.client.AMQSession.dispatch(AMQSession.java:3399)
        at org.apache.qpid.client.message.UnprocessedMessage.dispatch(UnprocessedMessage.java:54)
        at org.apache.qpid.client.AMQSession$Dispatcher.run(AMQSession.java:3547)
        at java.base/java.lang.Thread.run(Thread.java:829)
2021-12-08T14:31:37.297+0000 ERROR EvaluationSubscriber While attempting to recover a session for evaluation gK3v-6WzJtfsvFC1oz2lHokfTPY in subscriber SEWCDdrrnF-Umgp4qHSRG5KFA8M, encountered an error that prevented recovery.
javax.jms.IllegalStateException: Object org.apache.qpid.client.AMQSession_0_8@2280bd44 has been closed
        at org.apache.qpid.client.Closeable.checkNotClosed(Closeable.java:62)
        at org.apache.qpid.client.AMQSession.checkNotClosed(AMQSession.java:484)
        at org.apache.qpid.client.AMQSession.recover(AMQSession.java:1866)
        at wres.events.subscribe.EvaluationSubscriber.attemptRetry(EvaluationSubscriber.java:1189)
        at wres.events.subscribe.EvaluationSubscriber.recover(EvaluationSubscriber.java:1091)
        at wres.events.subscribe.EvaluationSubscriber.lambda$getEvaluationListener$3(EvaluationSubscriber.java:936)
        at org.apache.qpid.client.BasicMessageConsumer.notifyMessage(BasicMessageConsumer.java:758)
        at org.apache.qpid.client.BasicMessageConsumer.notifyMessage(BasicMessageConsumer.java:732)
        at org.apache.qpid.client.AMQSession$Dispatcher.notifyConsumer(AMQSession.java:3705)
        at org.apache.qpid.client.AMQSession$Dispatcher.dispatchMessage(AMQSession.java:3637)
        at org.apache.qpid.client.AMQSession$Dispatcher.access$1300(AMQSession.java:3406)
        at org.apache.qpid.client.AMQSession.dispatch(AMQSession.java:3399)
        at org.apache.qpid.client.message.UnprocessedMessage.dispatch(UnprocessedMessage.java:54)
        at org.apache.qpid.client.AMQSession$Dispatcher.run(AMQSession.java:3547)
        at java.base/java.lang.Thread.run(Thread.java:829)
2021-12-08T14:31:37.314+0000 INFO ProcessorHelper Wrote the following output: []
2021-12-08T14:31:37.835+0000 ERROR Main Operation 'execute' completed unsuccessfully
wres.pipeline.InternalWresException: Could not complete project execution
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:331)
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:183)
        at wres.MainFunctions.execute(MainFunctions.java:134)
        at wres.MainFunctions.call(MainFunctions.java:96)
        at wres.Main.main(Main.java:113)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'gK3v-6WzJtfsvFC1oz2lHokfTPY': 
        at wres.pipeline.ProcessorHelper.processEvaluation(ProcessorHelper.java:277)
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:306)
        ... 4 common frames omitted
Caused by: wres.pipeline.WresProcessingException: Project failed to complete with the following error: 
        at wres.pipeline.ProcessorHelper.processProjectConfig(ProcessorHelper.java:522)
        at wres.pipeline.ProcessorHelper.processEvaluation(ProcessorHelper.java:215)
        ... 5 common frames omitted
Caused by: java.lang.NullPointerException: null
        at wres.io.config.ConfigHelper.getConcreteFeatureDimension(ConfigHelper.java:585)
        at wres.io.thresholds.ExternalThresholdReader.readWRDSThresholds(ExternalThresholdReader.java:206)
        at wres.io.thresholds.ExternalThresholdReader.read(ExternalThresholdReader.java:84)
        at wres.io.thresholds.ThresholdReader.read(ThresholdReader.java:55)
        at wres.pipeline.ProcessorHelper.processProjectConfig(ProcessorHelper.java:429)
        ... 6 common frames omitted

```

Should probably shrink the declaration to only a couple features, since, as is, it takes a while for it to complete.

Hank

Adjusting system test scenario802 to use a `feature_name_from` of `baseline`, it Looks like the OP has been fixed, although there are some typos in the message, i.e., `threshold_service` --> `threshold_sources`:

```
wres.pipeline.UserInputException: The evaluation declaration was invalid.
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:166)
	at wres.Functions.evaluate(Functions.java:135)
	at wres.Functions.evaluate(Functions.java:97)
	at wres.helpers.MainUtilities.call(MainUtilities.java:107)
	at wres.Main.completeExecution(Main.java:206)
	at wres.Main.main(Main.java:126)
	at wres.MainLocal.main(MainLocal.java:805)
Caused by: wres.config.yaml.DeclarationException: Encountered 2 error(s) in the declared evaluation, which must be fixed:
    - The 'threshold_service' declaration requested that feature names with an orientation of 'baseline' are used to correlate features with thresholds, but no 'baseline' dataset was discovered. Please add a 'baseline' dataset or fix the 'feature_name_from' in the 'threshold_service' declaration.
    - Discovered declaration for a 'threshold_service', which requests thresholds whose feature names have an orientation of 'baseline'. However, 1 feature(s) were discovered with a missing 'baseline' feature name. Please fix the 'feature_name_from' in the 'threshold_service' declaration or supply fully composed feature tuples with an appropriate feature for the 'baseline' dataset. The first ten (or fewer) feature tuples with a missing baseline feature were: [{"left":{"name":"CKLN6"},"right":{"name":"CKLN6SUQ"}}]
	at wres.config.yaml.DeclarationValidator.notify(DeclarationValidator.java:483)
	at wres.config.yaml.DeclarationValidator.validate(DeclarationValidator.java:314)
	at wres.config.yaml.DeclarationFactory.from(DeclarationFactory.java:267)
	at wres.config.MultiDeclarationFactory.from(MultiDeclarationFactory.java:92)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:161)
	... 6 common frames omitted
```

Will fix those typos.

Fixed the typos.

```
wres.pipeline.UserInputException: The evaluation declaration was invalid.
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:166)
	at wres.Functions.evaluate(Functions.java:135)
	at wres.Functions.evaluate(Functions.java:97)
	at wres.helpers.MainUtilities.call(MainUtilities.java:107)
	at wres.Main.completeExecution(Main.java:206)
	at wres.Main.main(Main.java:126)
	at wres.MainLocal.main(MainLocal.java:805)
Caused by: wres.config.yaml.DeclarationException: Encountered 2 error(s) in the declared evaluation, which must be fixed:
    - The 'threshold_sources' declaration requested that feature names with an orientation of 'baseline' are used to correlate features with thresholds, but no 'baseline' dataset was discovered. Please add a 'baseline' dataset or fix the 'feature_name_from' in the 'threshold_sources' declaration.
    - Discovered declaration of 'threshold_sources', which requests thresholds whose feature names have an orientation of 'baseline'. However, 1 feature(s) were discovered with a missing 'baseline' feature name. Please fix the 'feature_name_from' in the 'threshold_sources' declaration or supply fully composed feature tuples with an appropriate feature for the 'baseline' dataset. The first ten (or fewer) feature tuples with a missing baseline feature were: [{"left":{"name":"CKLN6"},"right":{"name":"CKLN6SUQ"}}]
	at wres.config.yaml.DeclarationValidator.notify(DeclarationValidator.java:483)
	at wres.config.yaml.DeclarationValidator.validate(DeclarationValidator.java:314)
	at wres.config.yaml.DeclarationFactory.from(DeclarationFactory.java:267)
	at wres.config.MultiDeclarationFactory.from(MultiDeclarationFactory.java:92)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:161)
```

```
wres.pipeline.UserInputException: The evaluation declaration was invalid.
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:166)
	at wres.Functions.evaluate(Functions.java:135)
	at wres.Functions.evaluate(Functions.java:97)
	at wres.helpers.MainUtilities.call(MainUtilities.java:107)
	at wres.Main.completeExecution(Main.java:206)
	at wres.Main.main(Main.java:126)
	at wres.MainLocal.main(MainLocal.java:805)
Caused by: wres.config.yaml.DeclarationException: Encountered 1 error(s) in the declared evaluation, which must be fixed:
    - Discovered one or more thresholds with a 'feature_name_from' of 'baseline', but the evaluation does not declare a 'baseline' dataset. Please fix the 'feature_name_from' or declare a 'baseline' dataset and try again.
	at wres.config.yaml.DeclarationValidator.notify(DeclarationValidator.java:485)
	at wres.config.yaml.DeclarationValidator.validate(DeclarationValidator.java:316)
	at wres.config.yaml.DeclarationFactory.from(DeclarationFactory.java:267)
	at wres.config.MultiDeclarationFactory.from(MultiDeclarationFactory.java:92)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:161)
	... 6 common frames omitted
```

--- a/wres-config/src/wres/config/yaml/DeclarationValidator.java
+++ b/wres-config/src/wres/config/yaml/DeclarationValidator.java
@@ -122,7 +122,9 @@ public class DeclarationValidator
     private static final String THE_EVALUATION_REQUESTED_THE_SAMPLING_UNCERTAINTY =
             "The evaluation requested the 'sampling_uncertainty' ";
     /** Re-used string. */
-    private static final String AND_TRY_AGAIN = "and try again.";
+    private static final String TRY_AGAIN = "try again.";
+    /** Re-used string. */
+    private static final String AND_TRY_AGAIN = "and " + TRY_AGAIN;
 
     /**
      * Performs validation against the schema, followed by "business-logic" validation if there are no schema
@@ -1246,7 +1248,7 @@ public class DeclarationValidator
                                                              + ( variables.size() - named.size() )
                                                              + ". Please fix the 'minimum' and/or "
                                                              + "'maximum' value associated with these covariates and "
-                                                             + "try again." )
+                                                             + TRY_AGAIN )
                                            .build();
             events.add( event );
         }
@@ -1542,7 +1544,7 @@ public class DeclarationValidator
                                                                           .getFunction()
                                                              + "'. If this is incorrect, please add the correct "
                                                              + "'rescale_function' for the covariate dataset and "
-                                                             + "try again." )
+                                                             + TRY_AGAIN )
                                            .build();
             events.add( event );
         }
@@ -2365,8 +2367,14 @@ public class DeclarationValidator
         List<EvaluationStatusEvent> events =
                 new ArrayList<>( DeclarationValidator.thresholdSourcesAreValid( declaration ) );
 
+        // Check that value thresholds include units
         List<EvaluationStatusEvent> valueThresholds = DeclarationValidator.valueThresholdsIncludeUnits( declaration );
         events.addAll( valueThresholds );
+
+        // Check that the feature orientation is consistent with other declaration
+        List<EvaluationStatusEvent> featureThresholds =
+                DeclarationValidator.thresholdFeatureNameFromIsValid( declaration );
+        events.addAll( featureThresholds );
 
         return Collections.unmodifiableList( events );
     }
@@ -2409,6 +2417,36 @@ public class DeclarationValidator
                                            .setEventMessage( "Discovered one or more real-valued thresholds without a "
                                                              + "declared threshold unit. "
                                                              + extra )
+                                           .build();
+            events.add( event );
+        }
+
+        return Collections.unmodifiableList( events );
+    }
+
+    /**
+     * Checks that the {@code feature_name_from} declaration associated with thresholds is valid, specifically that a
+     * baseline dataset exists when a baseline orientation is declared.
+     * @param declaration the declaration
+     * @return the validation events encountered
+     */
+    private static List<EvaluationStatusEvent> thresholdFeatureNameFromIsValid( EvaluationDeclaration declaration )
+    {
+        List<EvaluationStatusEvent> events = new ArrayList<>();
+
+        Set<Threshold> thresholds = DeclarationUtilities.getInbandThresholds( declaration );
+        if ( thresholds.stream()
+                       .anyMatch( t -> t.featureNameFrom() == DatasetOrientation.BASELINE )
+             && !DeclarationUtilities.hasBaseline( declaration ) )
+        {
+            EvaluationStatusEvent event
+                    = EvaluationStatusEvent.newBuilder()
+                                           .setStatusLevel( StatusLevel.ERROR )
+                                           .setEventMessage( "Discovered one or more thresholds with a "
+                                                             + "'feature_name_from' of 'baseline', but the evaluation "
+                                                             + "does not declare a 'baseline' dataset. Please fix the "
+                                                             + "'feature_name_from' or declare a 'baseline' dataset and "
+                                                             + TRY_AGAIN )
                                            .build();
             events.add( event );
         }
@@ -2690,13 +2728,13 @@ public class DeclarationValidator
             EvaluationStatusEvent event
                     = EvaluationStatusEvent.newBuilder()
                                            .setStatusLevel( StatusLevel.ERROR )
-                                           .setEventMessage( "The 'threshold_service' declaration requested that "
+                                           .setEventMessage( "The 'threshold_sources' declaration requested that "
                                                              + "feature names with an orientation of '"
                                                              + DatasetOrientation.BASELINE
                                                              + "' are used to correlate features with thresholds, but "
                                                              + "no 'baseline' dataset was discovered. Please add a "
                                                              + "'baseline' dataset or fix the 'feature_name_from' "
-                                                             + "in the 'threshold_service' declaration." )
+                                                             + "in the 'threshold_sources' declaration." )
                                            .build();
 
             events.add( event );
@@ -2714,7 +2752,7 @@ public class DeclarationValidator
             EvaluationStatusEvent event
                     = EvaluationStatusEvent.newBuilder()
                                            .setStatusLevel( StatusLevel.ERROR )
-                                           .setEventMessage( "Discovered declaration for a 'threshold_service', which "
+                                           .setEventMessage( "Discovered declaration of 'threshold_sources', which "
                                                              + "requests thresholds whose feature names have an "
                                                              + "orientation of '"
                                                              + DatasetOrientation.BASELINE
@@ -2723,7 +2761,7 @@ public class DeclarationValidator
                                                              + " feature(s) were discovered with a missing '"
                                                              + DatasetOrientation.BASELINE
                                                              + "' feature name. Please fix the 'feature_name_from' in "
-                                                             + "the 'threshold_service' declaration or supply fully "
+                                                             + "the 'threshold_sources' declaration or supply fully "
                                                              + "composed feature tuples with an appropriate feature "
                                                              + "for the '"
                                                              + DatasetOrientation.BASELINE

--- a/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
+++ b/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
@@ -1615,6 +1615,36 @@ class DeclarationValidatorTest
     }
 
     @Test
+    void testFeaturefulThresholdsWithFeatureNameFromBaselineAndNoBaselineProducesError()
+    {
+        Geometry featureFoo = Geometry.newBuilder()
+                                      .setName( "foo" )
+                                      .build();
+        Threshold one = Threshold.newBuilder()
+                                 .setLeftThresholdValue( 1.0 )
+                                 .build();
+        wres.config.yaml.components.Threshold wrappedOne = ThresholdBuilder.builder()
+                                                                           .threshold( one )
+                                                                           .feature( featureFoo )
+                                                                           .featureNameFrom( DatasetOrientation.BASELINE )
+                                                                           .type( ThresholdType.VALUE )
+                                                                           .build();
+        EvaluationDeclaration declaration =
+                EvaluationDeclarationBuilder.builder()
+                                            .left( this.defaultDataset )
+                                            .right( this.defaultDataset )
+                                            .thresholds( Set.of( wrappedOne ) )
+                                            .build();
+
+        List<EvaluationStatusEvent> events = DeclarationValidator.validate( declaration );
+
+        assertTrue( DeclarationValidatorTest.contains( events,
+                                                       "Please fix the 'feature_name_from' or declare a "
+                                                       + "'baseline' dataset",
+                                                       StatusLevel.ERROR ) );
+    }
+
+    @Test
     void testFeaturefulThresholdsAndNoFeaturesProducesWarning()
     {
         Geometry featureFoo = Geometry.newBuilder()

